packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2929,303 +3020,326 @@ class GetLocalDiagCopy {
-#ifdef HAVE_TPETRA_DEBUG
-          std::ostream& err = this->markLocalErrorAndGetStream ();
-          err << prefix << "At i = " << i << ", GID " << gidsOut(k)
-              << " is not owned by the calling process.";
-#endif // HAVE_TPETRA_DEBUG
-          continue;
-        }
-      }
-
-      // Combine the incoming data with the matrix's current data.
-      LO numCombd = 0;
-      const LO* const lidsRaw = const_cast<const LO*> (lidsOut.data ());
-      const Scalar* const valsRaw =
-        reinterpret_cast<const Scalar*> (const_cast<const ST*> (valsOut.data ()));
-      if (CM == ADD) {
-        numCombd = this->sumIntoLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
-      } else if (CM == INSERT || CM == REPLACE) {
-        numCombd = this->replaceLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
-      } else if (CM == ABSMAX) {
-        numCombd = this->absMaxLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
-      }
+      const auto policy = Kokkos::RangePolicy<host_exec>(0, numImportLIDs+1);
+      Kokkos::parallel_scan
+        (policy,
+         [=](const int &i, size_t &update, const bool &final) {
+          if (final) offset(i) = update;
+          update += (i == numImportLIDs ? 0 : numPacketsPerLIDHost(i));
+        });
+    }
+
+    // this variable does not matter with a race condition (just error flag)
+    Kokkos::View<bool,host_exec,Kokkos::MemoryTraits<Kokkos::Atomic> > errorDuringUnpack("errorDuringUnpack");
+    errorDuringUnpack() = false;
+    {
+      typedef Kokkos::TeamPolicy<host_exec> policy_type;
+      const auto policy = 
+        policy_type(numImportLIDs, 1, 1)
+        .set_scratch_size(0, Kokkos::PerTeam(sizeof(GO(0))*maxRowNumEnt+
+                                             sizeof(LO(0))*maxRowNumEnt+
+                                             sizeof(impl_scalar_type(0))*maxRowNumScalarEnt));
+      
+      typedef typename host_exec::scratch_memory_space host_scratch_space;
+      Kokkos::parallel_for
+        (policy, 
+         [=](const typename policy_type::member_type &member) {
+          const int i = member.league_rank();            
+          Kokkos::View<GO*,host_scratch_space> gblColInds(member.team_scratch(0), maxRowNumEnt);
+          Kokkos::View<LO*,host_scratch_space> lclColInds(member.team_scratch(0), maxRowNumEnt);
+          Kokkos::View<impl_scalar_type*,host_scratch_space> vals(member.team_scratch(0), maxRowNumScalarEnt);
+
+          const size_t offval = offset(i);
+          const LO lclRow = importLIDsHost(i);
+          const size_t numBytes = numPacketsPerLIDHost(i);
+          const size_t numEnt =
+            unpackRowCount<impl_scalar_type, LO, GO, host_exec>
+            (importsByteHost, offval, numBytes, numBytesPerValue);
+
+          if (numBytes > 0) {
+            if (numEnt > maxRowNumEnt) {
+              errorDuringUnpack() = true;
+              if (debug) {
+                std::ostream& err = this->markLocalErrorAndGetStream ();
```

Comment:
Local error is a `Teuchos::RCP<bool>`, which is not thread safe by default.  It would be safer just to set `errorDuringUnpack()` for now inside the parallel loop, then call `markLocalErrorAndGetStream()` outside.